### ISSUE TITLE
Jj diff inline change indicators highlighting bytes not grapheme clusters

### ISSUE THREAD

**Issue by xelxebar** [martinvonz/jj#1381]
### Overview

When a line change contains multi-byte characters, `jj diff` highlights the changed bytes instead of the changed Unicode extended grapheme clusters (i.e. "visual characters").

### Reproduction

```
$ mkdir diff-test
$ cd diff-test/
$ jj init --git
Initialized repo in "."
$ echo '⊢' >file
$ jj commit -m 'file'
Working copy now at: ab1307ab0db8 (no description set)
$ echo '⊣' >file
$ jj diff
Modified regular file file:
   1    1: <E2><8A><A2><A3>
```

where `<E2><8A>` is highlighted in white, `<A2>` in red, and `<A3>` in green. Note that the byte sequences correspond to the UTF-8 encodings of `⊢` (U+22A2) and `⊣` (U+22A3).

### Expected Output

Of course, this is what we expect:

```
$ jj diff
Modified regular file file:
   1    1: ⊢⊣
```

where `⊢` is highlighted in red and `⊣` in green.

**Comment by martinvonz** [comment]
Makes sense. That's probably a fix around here: https://github.com/martinvonz/jj/blob/e64ca31bfda16d41a4c7d9cd683b3c11d7e5ae80/lib/src/diff.rs#L45-L78

Do you feel like fixing that?

**Comment by xelxebar** [comment]
I've never Rusted before but can give it a shot.

**Comment by xelxebar** [comment]
Okay, I am just winging this for now, without any proper study, but have hit somewhat of a roadblock. Would you mind prodding me in the right direction?

I replaced the `text.iter()` calls with a function from the `unicode-segmentation` crate gives an iterator over (extended) grapheme clusters. Specifically,  `grapheme_indices` iterates over graphemes and their byte offests. So far so good, if I am understanding correctly.

However, graphemes are represented as `str`s, so now `is_word_grapheme` needs to take a `str` instead of `u8`, but the type checker is yelling incomprehensible things at me:

```
77 |         if !is_word_grapheme(*g) {
   |                              ^^ doesn't have a size known at compile-time
```
 
I mean, obviously, but I'm not sure what to do about it. Patch of changes to `diff.rs` below. Any direction here would be much appreciated!

```diff
diff --git a/lib/src/diff.rs b/lib/src/diff.rs
index aefdd0a656...c59a1eec20 100644
--- a/lib/src/diff.rs
+++ b/lib/src/diff.rs
@@ -12,16 +12,19 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+extern crate unicode_segmentation;
+
 use std::cmp::{max, min, Ordering};
 use std::collections::{BTreeMap, HashMap};
 use std::fmt::{Debug, Formatter};
 use std::ops::Range;
 use std::slice;
-
 use itertools::Itertools;
 
 use crate::nightly_shims::BTreeMapExt;
 
+use unicode_segmentation::UnicodeSegmentation;
+
 pub fn find_line_ranges(text: &[u8]) -> Vec<Range<usize>> {
     let mut ranges = vec![];
     let mut start = 0;
@@ -42,21 +45,22 @@
     ranges
 }
 
-fn is_word_byte(b: u8) -> bool {
+fn is_word_grapheme(g: str) -> bool {
     // TODO: Make this configurable (probably higher up in the call stack)
-    matches!(b, b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_')
-}
+    matches!(g, b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'_')
 
+}
+
 pub fn find_word_ranges(text: &[u8]) -> Vec<Range<usize>> {
     let mut word_ranges = vec![];
     let mut word_start_pos = 0;
     let mut in_word = false;
-    for (i, b) in text.iter().enumerate() {
-        if in_word && !is_word_byte(*b) {
+    for (i, g) in String::from_utf8_lossy(text).grapheme_indices(true).enumerate() {
+        if in_word && !is_word_grapheme(*g) {
             in_word = false;
             word_ranges.push(word_start_pos..i);
             word_start_pos = i;
-        } else if !in_word && is_word_byte(*b) {
+        } else if !in_word && is_word_grapheme(*g) {
             in_word = true;
             word_start_pos = i;
         }
@@ -69,8 +73,8 @@
 
 pub fn find_nonword_ranges(text: &[u8]) -> Vec<Range<usize>> {
     let mut ranges = vec![];
-    for (i, b) in text.iter().enumerate() {
-        if !is_word_byte(*b) {
+    for (i, g) in String::from_utf8_lossy(text).grapheme_indices(true).enumerate() {
+        if !is_word_grapheme(*g) {
             ranges.push(i..i + 1);
         }
     }

```

**Comment by martinvonz** [comment]
You want to make that `g: &str`, i.e. a reference to the `str`.

Thanks for working on this!

**Comment by xelxebar** [comment]
That did the trick. Here's a PR, for detailed discussions https://github.com/martinvonz/jj/pull/1391.

**Comment by yuja** [comment]
Should be fixed by #1391.
